L1Trigger/Phase2L1ParticleFlow/src/ref/pfalgo2hgc_ref.cpp

```diff
@@ -0,0 +1,196 @@
+#include "pfalgo2hgc_ref.h"
+
+#ifndef CMSSW_GIT_HASH
+#include "../DiscretePFInputs.h"
+#else
+#include "../../interface/DiscretePFInputs.h"
+#endif
+
+#include "../utils/Firmware2DiscretePF.h"
+#include <cmath>
+#include <cstdio>
+#include <algorithm>
+#include <memory>
+
+int g_pfalgo2hgc_debug_ref_ = 0;
+
+void pfalgo2hgc_ref_set_debug(int debug) { g_pfalgo2hgc_debug_ref_ = debug; }
+
+void pfalgo2hgc_ref(const pfalgo_config &cfg,
+                    const HadCaloObj calo[/*cfg.nCALO*/],
+                    const TkObj track[/*cfg.nTRACK*/],
+                    const MuObj mu[/*cfg.nMU*/],
+                    PFChargedObj outch[/*cfg.nTRACK*/],
+                    PFNeutralObj outne[/*cfg.nSELCALO*/],
+                    PFChargedObj outmu[/*cfg.nMU*/]) {
+  if (g_pfalgo2hgc_debug_ref_) {
+#ifdef L1Trigger_Phase2L1ParticleFlow_DiscretePFInputs_MORE
+    for (unsigned int i = 0; i < cfg.nTRACK; ++i) {
+      if (track[i].hwPt == 0)
+        continue;
+      l1tpf_impl::PropagatedTrack tk;
+      fw2dpf::convert(track[i], tk);
+      printf(
+          "FW  \t track %3d: pt %8d [ %7.2f ]  calo eta %+7d [ %+5.2f ]  calo phi %+7d [ %+5.2f ]  calo ptErr %6d [ "
+          "%7.2f ]   tight %d\n",
+          i,
+          tk.hwPt,
+          tk.floatPt(),
+          tk.hwEta,
+          tk.floatEta(),
+          tk.hwPhi,
+          tk.floatPhi(),
+          tk.hwCaloPtErr,
+          tk.floatCaloPtErr(),
+          int(track[i].hwTightQuality));
+    }
+    for (unsigned int i = 0; i < cfg.nCALO; ++i) {
+      if (calo[i].hwPt == 0)
+        continue;
+      l1tpf_impl::CaloCluster c;
+      fw2dpf::convert(calo[i], c);
+      printf(
+          "FW  \t calo  %3d: pt %8d [ %7.2f ]  calo eta %+7d [ %+5.2f ]  calo phi %+7d [ %+5.2f ]  calo emPt %7d [ "
+          "%7.2f ]   isEM %d \n",
+          i,
+          c.hwPt,
+          c.floatPt(),
+          c.hwEta,
+          c.floatEta(),
+          c.hwPhi,
+          c.floatPhi(),
+          c.hwEmPt,
+          c.floatEmPt(),
+          c.isEM);
+    }
+    for (unsigned int i = 0; i < cfg.nMU; ++i) {
+      if (mu[i].hwPt == 0)
+        continue;
+      l1tpf_impl::Muon muon;
+      fw2dpf::convert(mu[i], muon);
+      printf("FW  \t muon  %3d: pt %8d [ %7.2f ]  muon eta %+7d [ %+5.2f ]  muon phi %+7d [ %+5.2f ]   \n",
+             i,
+             muon.hwPt,
+             muon.floatPt(),
+             muon.hwEta,
+             muon.floatEta(),
+             muon.hwPhi,
+             muon.floatPhi());
+    }
+#endif
+  }
+
+  // constants
+  const pt_t TKPT_MAX_LOOSE = cfg.tk_MAXINVPT_LOOSE;
+  const pt_t TKPT_MAX_TIGHT = cfg.tk_MAXINVPT_TIGHT;
+  const int DR2MAX = cfg.dR2MAX_TK_CALO;
+
+  ////////////////////////////////////////////////////
+  // TK-MU Linking
+  std::unique_ptr<bool[]> isMu(new bool[cfg.nTRACK]);
```

Comment:
why not a vector?

Reply:
I don't know why the choice was made.

On the other hand, a `unique_ptr<T[]>` is two pointers smaller than a `vector<T>`, and does not offer the interface to push_back, resize, etc., which is not needed here.

Reply:
because `vector<bool>` is not a good vector

Reply:
What impact would the `vector<bool>` specialization have in practice?